### ISSUE TITLE
Document other kinds of unstability in The Unstable Book

### ISSUE THREAD

**Issue by frewsxcv** [rust-lang/rust#41142]
Right now, The Unstable Book is just a list of unstable lang/lib features. There are other things like compiler flags which are also considered unstable that should be documented in The Unstable Book. This requires adjusting [the tidy lint](https://github.com/rust-lang/rust/blob/master/src/tools/tidy/src/unstable_book.rs) a bit to accommodate other sections. Right now I'm thinking that all sections (features, flags, etc) will live in their own subdirectory.

**Comment by frewsxcv** [comment]
Relevant thread: https://github.com/rust-lang/rust/pull/40018#issuecomment-292421507

**Comment by frewsxcv** [comment]
Reminder for future self: remove this hack e280515499dd24dcfbfe9455806f328b879b243d

**Comment by brson** [comment]
This may not be the issue, but there are a number of stable features that are actually just "stable", like target specs and llvm plugins, where we don't even try to maintain stability. It would be nice to have a discussion of those somewhere.

**Comment by frewsxcv** [comment]
I'm waiting on https://github.com/rust-lang/rust/pull/40018 to merge before I work on this.

**Comment by frewsxcv** [comment]
Copying @est31's text from https://github.com/rust-lang/rust/issues/41289: 

> I think for greater categorisation, the unstable book should be divided into sections.
> 
> A proposal:
> 
> * Compiler internal lang features ("not intended for general use")
> * Permanently unstable features that are not compiler internal (like the llvm intrinsics)
> * Novel lang features (with intent to stabilize them some day)
> * Library features
> * ??
> 
> Some sections should have a higher standard of documentation, like the novel lang features, others like the library features already have documentation in rustdoc (see #41288), so their documentation should be limited to a link to the tracking issue (and maybe their rustdoc), and the "not intended for general use" features can get a lower priority to be documented as well.

**Comment by frewsxcv** [comment]
This was partially completed in https://github.com/rust-lang/rust/issues/41295.

**Comment by steveklabnik** [comment]
In https://github.com/rust-lang/rust/issues/41288 @est31 suggests that we should keep the library features minimal; I agree. They should exist, so you can see everything that's unstable, but there's no reason to duplicate their docs.

**Comment by steveklabnik** [comment]
```
19:39 < frewsxcv> i think the main thing missing for this is all the unstable
                  command line options
```

**Comment by frewsxcv** [comment]
regarding unsafe command line options, one can see them via `rustc -Z help`. it looks like the options are all defined [in this `src/librustc/session/config.rs` file](https://github.com/rust-lang/rust/blob/ae8efdc87d5f4e22e05e1b60a72272cee87fa74c/src/librustc/session/config.rs#L913). they seem to all be exposed on [`rustc::session::Session`](https://manishearth.github.io/rust-internals-docs/rustc/session/struct.Session.html) and [`rustc::session::config::DebuggingOptions`](https://manishearth.github.io/rust-internals-docs/rustc/session/config/struct.DebuggingOptions.html#structfield.time_llvm_passes). seems like we should be able to autogenerate all the unsafe command line options pages like we do for the other unstable book sections (via `src/tools/unstable-book-gen`)

**Comment by steveklabnik** [comment]
Triage: we now have https://doc.rust-lang.org/rustc/command-line-arguments.html, but it doesn't talk about `-Z`.

**Comment by DevQps** [comment]
I just saw that the Unstable Book already contains a list of unstable compiler flags [here](https://doc.rust-lang.org/unstable-book/compiler-flags.html) which I guess means that this issue is already solved right? 

Maybe we should create a separate issue for the missing -Z description in https://doc.rust-lang.org/rustc/command-line-arguments.html and close this one? Maybe also create a separate issue to autogenerate the unsafe nightly commands for the Unstable Book. That way we can keep the tasks that need to be done clearly isolated.

@steveklabnik If you agree I can create these new issues with a nice description of what needs to be done!

**Comment by steveklabnik** [comment]
> Maybe we should create a separate issue for the missing -Z description in https://doc.rust-lang.org/rustc/command-line-arguments.html and close this one?

Yes! Alternatively, if this is something you'd like to send a PR for, we can just keep this issue open and use that PR to close it.

>  Maybe also create a separate issue to autogenerate the unsafe nightly commands for the Unstable Book.

I would prefer not to, I don't think this is the job for the unstable book.

**Comment by DevQps** [comment]
> Yes! Alternatively, if this is something you'd like to send a PR for, we can just keep this issue open and use that PR to close it.
I'll go ahead and do that then!

> I would prefer not to, I don't think this is the job for the unstable book.
I can understand that you have this view! I do not really have much of a view myself on this, though I do think that we should keep duplication between the `rustc` docs and other docs to a minimum (or to keep documentation duplication to a minimum in general because its easier to maintain). The `rustc` docs are probably a better place to document this.

However, this does lead me to the following question: Currently, there is already a [compiler flags section](https://doc.rust-lang.org/unstable-book/compiler-flags.html) in the unstable book. For example the `-Z emit-stack-sizes` flag. I think we should either try to complete this section or remove/move it to the `rustc` docs all-together. What is your opinion on this?

**Comment by steveklabnik** [comment]
Moving it to the `rustc` docs seems appropriate to me!

**Comment by DevQps** [comment]
> Moving it to the `rustc` docs seems appropriate to me!

I created an issue for this here: #59679 
Then we can use this issue for adding -Z to the compiler flags section!
